libraries/tickfs/src/tickfs.rs

```diff
@@ -0,0 +1,673 @@
+    region_size: usize,
+    read_buffer: Cell<&'a mut [u8]>,
+    phantom_hasher: PhantomData<H>,
+}
+
+/// This is the current object header used for TickFS objects
+struct ObjectHeader {
+    version: u8,
+    // In reality this is a u4.
+    flags: u8,
+    // In reality this is a u12.
+    len: u16,
+    hashed_key: u64,
+}
+
+pub(crate) const FLAGS_VALID: u8 = 8;
+
+impl ObjectHeader {
+    fn new(hashed_key: u64, len: u16) -> Self {
+        assert!(len < 0xFFF);
+        Self {
+            version: VERSION,
+            flags: FLAGS_VALID,
+            len,
+            hashed_key,
+        }
+    }
+}
+
+// A list of offsets into the ObjectHeader
+pub(crate) const VERSION_OFFSET: usize = 0;
+pub(crate) const LEN_OFFSET: usize = 1;
+pub(crate) const HASH_OFFSET: usize = 3;
+pub(crate) const HEADER_LENGTH: usize = HASH_OFFSET + 8;
+pub(crate) const CHECK_SUM_LEN: usize = 8;
+
+const MAIN_KEY: &str = "tickfs-super-key";
+
+/// This is the main TickFS struct.
+impl<'a, C: FlashController, H: Hasher> TickFS<'a, C, H> {
+    /// Create a new struct
+    ///
+    /// `C`: An implementation of the `FlashController` trait
+    /// `H`: An implementation of a `core::hash::Hasher` trait. This MUST
+    ///      always return the same hash for the same input. That is the
+    ///      implementation can NOT change over time.
+    ///
+    /// `controller`: An new struct implementing `FlashController`
+    /// `flash_size`: The total size of the flash used for TickFS
+    /// `region_size`: The smallest size that can be erased in a single operation.
+    ///                This must be a multiple of the start address and `flash_size`
+    ///
+    /// If the specified region has not already been setup for TickFS
+    /// the entire region will be erased.
+    pub fn new(
+        controller: C,
+        hash_function: (&mut H, &mut H),
+        read_buffer: &'a mut [u8],
+        flash_size: usize,
+        region_size: usize,
+    ) -> Option<Self> {
+        let tickfs = Self {
+            controller,
+            flash_size,
+            region_size,
+            read_buffer: Cell::new(read_buffer),
+            phantom_hasher: PhantomData,
+        };
+
+        let mut buf: [u8; 0] = [0; 0];
+
+        match tickfs.get_key(hash_function.0, MAIN_KEY, &mut buf) {
+            Ok(()) => {}
+            Err(_e) => {
+                // Erase all regions
+                for r in 0..(flash_size / region_size) {
+                    if tickfs.controller.erase_region(r).is_err() {
+                        return None;
+                    }
+                }
+
+                // Save the main key
+                if tickfs.append_key(hash_function.1, MAIN_KEY, &buf).is_err() {
+                    return None;
+                }
+            }
+        };
+
+        Some(tickfs)
+    }
+
+    /// Generate the hash and region number from a key
+    fn get_hash_and_region(&self, hash_function: &mut H, key: &str) -> (u64, usize) {
```

Comment:
The discussion in https://github.com/tock/tock/issues/1692 seemed to conclude that having keys be `&[u8]` or `u32` values, instead of `&str` values, would be preferred. Can you elaborate on why you chose `&str` keys?

Reply:
I have changed it to `&[u8]`